[Cocoa] Add support for overriding DNS resolution for WPT

pvollan opened the pull request:
#### ae5d4ebf2175309275a974221a1b85ddf55facac
<pre>
[Cocoa] Add support for overriding DNS resolution for WPT
<a href="https://bugs.webkit.org/show_bug.cgi?id=261038">https://bugs.webkit.org/show_bug.cgi?id=261038</a>
<a href="https://rdar.apple.com/82927182">rdar://82927182</a>

Reviewed by Alex Christensen.

Patch by Matthew Finkel and Anne van Kesteren with minor modifications by Per Arne.

Some ports don&apos;t have support for resolving arbitrary domains, but
web-platform-tests rely on this capability. Currently, the Apple ports only
use localhost, and 127.0.0.1 as a workaround where it&apos;s necessary, but this
affects our test coverage. This patch adds a very basic DNS resolver and
introduces a new command-line argument for run-webkit-tests; this is disabled
by default.

In 254273@main, glib ports gained support for resolving arbitrary domains,
and this moves Apple ports closer to supporting that. We expect to enable
this by default, along with rebaselining, in a later commit.

Previous attempts were made in <a href="https://bugs.webkit.org/show_bug.cgi?id=245294">https://bugs.webkit.org/show_bug.cgi?id=245294</a>,
and 264076@main introduced part of that behavior for a different reason. In
some situations, name resolutions falls back on that SPI, as well.

The basic DNS resolver only resolves A records for names ending with
&quot;localhost&quot; and for the host names defines by the port&apos;s localhost_aliases.
The server always returns 127.0.0.1 as the answer for those hosts. All other
queries return NXDomain. This resolver depends on the dnslib Python library.

On Apple ports, the iOS Simulator and Mac use different implementations. The
iOS simulator part is in the Network process because we need to define the
resolver in-process otherwise it won&apos;t be used. The way the resolver is
configured on Mac isn&apos;t supported on the simulator. On Mac, we use SPI and a
Network Extensions policy so this works when a VPN is enabled. We do this in
WebKitTestRunner so we don&apos;t need to add additional entitlements and weaken
the sandbox for the network process.

* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::platformInitializeNetworkProcessCocoa):
* Tools/Scripts/webkitpy/__init__.py:
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(parse_args):
* Tools/Scripts/webkitpy/layout_tests/servers/basic_dns_server.py: Added.
(Resolver):
(Resolver.__init__):
(Resolver.resolve):
* Tools/Scripts/webkitpy/layout_tests/servers/web_platform_test_server.py:
(WebPlatformTestServer.__init__):
(WebPlatformTestServer._spawn_process):
(WebPlatformTestServer._stop_running_server):
* Tools/Scripts/webkitpy/port/apple.py:
(ApplePort.__init__):
* Tools/Scripts/webkitpy/port/driver.py:
(Driver.cmd_line):
* Tools/Scripts/webkitpy/port/test.py:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunner-internal.entitlements:
* Tools/WebKitTestRunner/Options.cpp:
(WTR::handleOptionLocalDNSResolver):
(WTR::OptionsHandler::OptionsHandler):
* Tools/WebKitTestRunner/Options.h:
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaPlatformInitialize):
(WTR::TestController::cocoaDNSInitialize):

Canonical link: <a href="https://commits.webkit.org/300026@main">https://commits.webkit.org/300026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/283e818a676e9f08ba4d9217dfedb6638561fe0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40703 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31361 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127424 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73087 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/96bf5639-ce37-4f53-857a-26ed16206399) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49282 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91907 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61139 "Found 5 new test failures: imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub.html imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub.html imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub.html imported/w3c/web-platform-tests/resource-timing/resource_TAO_match_origin.htm imported/w3c/web-platform-tests/scroll-to-text-fragment/iframes.sub.html (failure)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bf37d380-6a7e-4dc8-b0b3-e67d2d5b5830) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123959 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108473 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72594 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d849352f-98fa-42e2-9dc1-eda531f7cee8) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/120364 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32086 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26577 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71013 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130279 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47934 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36418 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100516 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100418 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45825 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23875 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44622 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19207 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47792 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53505 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47263 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50610 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48947 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/WebKitTestRunner/Configurations/WebKitTestRunner-internal.entitlements:
```diff
@@ -8,6 +8,12 @@
 	</array>
 	<key>com.apple.private.webkit.adattributiond.testing</key>
 	<true/>
+	<key>com.apple.private.nehelper.privileged</key>
+	<true/>
+	<key>com.apple.private.neagent</key>
+	<true/>
+	<key>com.apple.networkd.persistent_interface</key>
+	<true/>
```
The commit message says:

> On Mac, we use SPI and a Network Extensions policy so this works when a VPN is enabled. We do this in WebKitTestRunner so we don't need to add additional entitlements and weaken the sandbox for the network process.

But given this is in `WebKitTestRunner-internal.entitlements` and not `WebKitTestRunner.entitlements` this means we only have this working when built with the internal SDK, does it not? Because if that's the case, then we're not actually succeeding at the goal of having WPT run with a full set of hostnames in WebKit CI.

pvollan replied:
That is a good point. I believe these entitlements are required for this to work when VPN is enabled, so we may still be able to achieve our goal of running with a full set of hostnames in WebKit CI.

Thanks for reviewing!

pvollan committed:
[Cocoa] Add support for overriding DNS resolution for WPT
https://bugs.webkit.org/show_bug.cgi?id=261038
rdar://82927182

Reviewed by Alex Christensen.

Patch by Matthew Finkel and Anne van Kesteren with minor modifications by Per Arne.

Some ports don't have support for resolving arbitrary domains, but
web-platform-tests rely on this capability. Currently, the Apple ports only
use localhost, and 127.0.0.1 as a workaround where it's necessary, but this
affects our test coverage. This patch adds a very basic DNS resolver and
introduces a new command-line argument for run-webkit-tests; this is disabled
by default.

In 254273@main, glib ports gained support for resolving arbitrary domains,
and this moves Apple ports closer to supporting that. We expect to enable
this by default, along with rebaselining, in a later commit.

Previous attempts were made in https://bugs.webkit.org/show_bug.cgi?id=245294,
and 264076@main introduced part of that behavior for a different reason. In
some situations, name resolutions falls back on that SPI, as well.

The basic DNS resolver only resolves A records for names ending with
"localhost" and for the host names defines by the port's localhost_aliases.
The server always returns 127.0.0.1 as the answer for those hosts. All other
queries return NXDomain. This resolver depends on the dnslib Python library.

On Apple ports, the iOS Simulator and Mac use different implementations. The
iOS simulator part is in the Network process because we need to define the
resolver in-process otherwise it won't be used. The way the resolver is
configured on Mac isn't supported on the simulator. On Mac, we use SPI and a
Network Extensions policy so this works when a VPN is enabled. We do this in
WebKitTestRunner so we don't need to add additional entitlements and weaken
the sandbox for the network process.

* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::platformInitializeNetworkProcessCocoa):
* Tools/Scripts/webkitpy/__init__.py:
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(parse_args):
* Tools/Scripts/webkitpy/layout_tests/servers/basic_dns_server.py: Added.
(Resolver):
(Resolver.__init__):
(Resolver.resolve):
* Tools/Scripts/webkitpy/layout_tests/servers/web_platform_test_server.py:
(WebPlatformTestServer.__init__):
(WebPlatformTestServer._spawn_process):
(WebPlatformTestServer._stop_running_server):
* Tools/Scripts/webkitpy/port/apple.py:
(ApplePort.__init__):
* Tools/Scripts/webkitpy/port/driver.py:
(Driver.cmd_line):
* Tools/Scripts/webkitpy/port/test.py:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunner-internal.entitlements:
* Tools/WebKitTestRunner/Options.cpp:
(WTR::handleOptionLocalDNSResolver):
(WTR::OptionsHandler::OptionsHandler):
* Tools/WebKitTestRunner/Options.h:
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaPlatformInitialize):
(WTR::TestController::cocoaDNSInitialize):

Canonical link: https://commits.webkit.org/300026@main
